File: app/src/main/java/io/oldering/tvfoot/red/matches/MatchesRepository.java
package io.oldering.tvfoot.red.matches;

import io.oldering.tvfoot.red.data.api.MatchService;
import io.oldering.tvfoot.red.data.entity.search.Filter;
import io.reactivex.Observable;

import static io.oldering.tvfoot.red.matches.MatchesViewState.Status.FIRST_PAGE_ERROR;
import static io.oldering.tvfoot.red.matches.MatchesViewState.Status.FIRST_PAGE_LOADED;
import static io.oldering.tvfoot.red.matches.MatchesViewState.Status.FIRST_PAGE_LOADING;
import static io.oldering.tvfoot.red.matches.MatchesViewState.Status.NEXT_PAGE_ERROR;
import static io.oldering.tvfoot.red.matches.MatchesViewState.Status.NEXT_PAGE_LOADED;
import static io.oldering.tvfoot.red.matches.MatchesViewState.Status.NEXT_PAGE_LOADING;

public class MatchesRepository {
  final MatchService matchService;
  private int pageIndex = 0;

  public MatchesRepository(MatchService matchService) {
    this.matchService = matchService;
  }

  public Observable<MatchesViewState> loadFirstPage() {
    return matchService.findFuture(Filter.builder().limit(30).offset(0).build())
        .toObservable()
        .map(MatchRowDisplayable::fromMatches)
        .map(matches -> MatchesViewState.builder()
            .matches(matches)
            .status(FIRST_PAGE_LOADED)
            .build())
        .startWith(MatchesViewState.builder().status(FIRST_PAGE_LOADING).build())
        .onErrorReturn(error -> MatchesViewState.builder()
            .firstPageError(error)
            .status(FIRST_PAGE_ERROR)
            .build());
  }

  public Observable<MatchesViewState> loadNextPage() {
    return matchService.findFuture(Filter.builder().limit(30).offset(30 * ++pageIndex).build())
        .toObservable()
        .map(MatchRowDisplayable::fromMatches)
        .map(
            matches -> MatchesViewState.builder().matches(matches).status(NEXT_PAGE_LOADED).build())
        .startWith(MatchesViewState.builder().status(NEXT_PAGE_LOADING).build())
        .onErrorReturn(error -> MatchesViewState.builder()
            .firstPageError(error)
            .status(NEXT_PAGE_ERROR)
            .build());
  }
}

File: app/src/main/java/com/benoitquenaudon/tvfoot/red/app/domain/matches/state/MatchesRepository.kt
package com.benoitquenaudon.tvfoot.red.app.domain.matches.state

import com.benoitquenaudon.tvfoot.red.api.TvfootService
import com.benoitquenaudon.tvfoot.red.app.data.entity.Match
import com.benoitquenaudon.tvfoot.red.app.data.entity.search.Filter
import io.reactivex.Single
import javax.inject.Inject

class MatchesRepository @Inject constructor(
    private val tvfootService: TvfootService
) : BaseMatchesRepository {
  override fun loadPage(pageIndex: Int): Single<List<Match>> {
    return tvfootService.findFuture(
        Filter(limit = MATCH_PER_PAGE, offset = MATCH_PER_PAGE * pageIndex))
  }

  companion object Constant {
    private val MATCH_PER_PAGE = 30
  }
}
